rfd/0008-application-access.md

```diff
@@ -0,0 +1,272 @@
+```
+
+The `proxy_service` section has been updated to support loading multiple certificate and key pairs. This allows loading per-application TLS certificates as well as loading of wildcard TLS certificates.
+
+```yaml
+  # List of certificates for the proxy to load.
+  https_keypairs:
+    - key_file: /var/lib/teleport/certs/proxy.example.com+3-key.pem
+      cert_file: /var/lib/teleport/certs/proxy.example.com+3.pem
+    - key_file: /var/lib/teleport/certs/app.example.com+3-key.pem
+      cert_file: /var/lib/teleport/certs/app.example.com+3.pem
+```
+
+An `app_service` section has been added to configure the application proxy service.
+
+```yaml
+# An application service establishes a reverse tunnel to the proxy which
+# is used both to heartbeat the presence of the application as well as
+# used to establish connections through.
+app_service:
+   # Enabled controls if the application service is enabled.
+   enabled: true
+
+   # A list of applications that will be proxied.
+   apps:
+     # Name of the application.
+   - name: jenkins
+     # URI is the address the application being proxied can be reached at.
+     uri: http://localhost:8080
+     # Public address is used to override the default address the application
+     # is avaiable at.
+     public_addr: jenkins.acme.teleport.dev
+     # Insecure skip verify is used to disable server TLS certificate
+     # verification. Useful for internal applications using a self signed
+     # certificate.
+     insecure_skip_verify: true
+     # Rewriting rules that get applied to every request.
+     rewrite:
+        # Rewrite the "Location" header on redirect "30{1-8}"   
+        # responses replacing the host with the public address of
+        # this application.   
+        redirect:
+           - "localhost"
+           - "jenkins.internal.dev"
+     # Static labels to assign to this application.
+     labels:
+        key: value
+     # Dynamic labels to assign to this application.
+     commands:
+     - name: "arch"
+       command: ["/bin/uname", "-p"]
+       period: "1h0m0s"
+```
+
+#### Command Line Flags
+
+The `teleport` process has been updated to support three additional flags that can be used to configure an application service in addition to the existing `--roles` and `--labels` flag (which now apply to the application, not the service).
+
+```
+$ teleport start \
+   # Define the role of service, in this case "app_service".
+   --roles="app" \
+   # Define the token the service will use to join the cluster.
+   --token=D83721493A6BE34862FAD8C9FFBDADA7 \
+   # Define any static or dynamic labels to apply to the application.
+   --labels="foo=bar,baz=qux" \
+   # Define the name of the application.
+   --app-name="jenkins" \
+   # Define the URI that the application is running at.
+   --app-uri="http://localhost:8080" \
+   # Define the public address, used to overide the automatically generated
+   # address of appName.proxyPublicAddr.
```

Comment:
Yes, the logic works as follows:

1. If you define `public_addr` it will be used.
2. If you do not define `public_addr`, Teleport will attempt to construct the `public_addr` from `appName.proxyPublicAddr`.
3. If you do not have `proxyPublicAddr` defined, Teleport will fall back to `appName.clusterName`.

The purpose of `public_addr` is to act as an override for the default that's generated for you if you want to run your application under some other domain.

Reply:
I'm planning on adding a section to this RFD explaining this by the way.